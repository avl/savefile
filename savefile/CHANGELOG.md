# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.17.2](https://github.com/avl/savefile/compare/savefile-v0.17.1...savefile-v0.17.2) - 2024-05-05

### Fixed
- Make comment clearer

## [0.17.1](https://github.com/avl/savefile/compare/savefile-v0.17.0...savefile-v0.17.1) - 2024-05-01

### Other
- Improve documentation

## [0.17.0-beta.15](https://github.com/avl/savefile/compare/savefile-v0.17.0-beta.14...savefile-v0.17.0-beta.15) - 2024-04-30

### Added
- Support for serializing 'str'

### Fixed
- Fix bug in deserializing old schemas

## [0.17.0-beta.14](https://github.com/avl/savefile/compare/savefile-v0.17.0-beta.13...savefile-v0.17.0-beta.14) - 2024-04-30

### Fixed
- Bad cycles detection

## [0.17.0-beta.13](https://github.com/avl/savefile/compare/savefile-v0.17.0-beta.12...savefile-v0.17.0-beta.13) - 2024-04-27

### Added
- savefile-abi2

These CHANGELOG-files are experimental.

## [0.17.0-beta.12](https://github.com/avl/savefile/compare/savefile-v0.17.0-beta.11...savefile-v0.17.0-beta.12) - 2024-04-27

### Other
- Unspecified work
