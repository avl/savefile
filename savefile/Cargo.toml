[package]
name = "savefile"
version = "0.17.3"
authors = ["Anders Musikka <anders@andersmusikka.se>"]
documentation = "https://docs.rs/savefile/"
homepage = "https://github.com/avl/savefile/"
repository = "https://github.com/avl/savefile"

exclude = [
    "*.bin"
]

description = "Simple, convenient, fast, versioned, binary serialization/deserialization library."

readme = "../README.md"

keywords = ["serialization", "deserialization", "introspection"]

categories = ["encoding"]

license = "MIT/Apache-2.0"

edition = "2021"

[features]
default = ["indexmap", "arrayvec", "smallvec", "bit-vec", "parking_lot","bit-set"]
bit-set = ["dep:bit-set","bit-vec"]
serde_derive = ["dep:serde_derive","serde"]

# Enable this to reduce risk of crashing on corrupt input. Provides sanity checks for sizes of objects.
# This is mostly to be able to run fuzzers against the deserializers without them being guaranteed to easily find out-of-memory crashes.
size_sanity_checks = []
# Use features only available on the nightly rust-compiler.
# Enabling this provides slightly better introspection support.
nightly=[]

compression = ["bzip2"]

encryption = ["ring", "rand"]

derive = ["dep:savefile-derive"]

[dependencies]
bit-vec = { version = "0.6", optional = true}
arrayvec = { version = "0.7", optional = true}
smallvec = { version = "1.11", optional = true}
indexmap = { version = "1.9", optional = true}
parking_lot = { version = "0.12", optional = true }
ring = {version = "0.16.9", optional = true}
rand = { version = "0.8", optional = true}
bzip2 = {version = "0.4.4", optional = true}
bit-set = {version = "0.5", optional = true}
rustc-hash = {version = "1.1", optional = true}
memoffset = "0.9"
byteorder = "1.4"
savefile-derive = {path="../savefile-derive", version = "=0.17.3", optional = true }
serde_derive = {version= "1.0", optional = true}
serde = {version= "1.0", optional = true}
quickcheck = {version= "1.0", optional = true}

[dev-dependencies]
savefile-derive = { path="../savefile-derive", version = "=0.17.3" }

[build-dependencies]
rustc_version="0.2"



