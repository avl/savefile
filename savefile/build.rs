extern crate rustc_version;
use rustc_version::{version_meta, Channel, Version};
fn main() {
    let version = version_meta().unwrap();
    if version.channel == Channel::Nightly {
        println!("cargo:rustc-cfg=feature=\"nightly\"");
    }
    if version.semver >= Version::new(1, 78, 0) {
        println!("cargo:rustc-cfg=feature=\"rust1_78\"");
    }
}
