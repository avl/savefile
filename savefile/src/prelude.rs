pub use {
    super::deserialize_slice_as_vec, super::get_schema, super::introspect_item, super::load, super::load_file,
    super::load_file_noschema, super::load_from_mem, super::load_noschema, super::save, super::save_file,
    super::save_file_noschema, super::save_noschema, super::save_to_mem, super::AbiRemoved, super::Canary1,
    super::Deserialize, super::Deserializer, super::Field, super::Introspect, super::IntrospectItem,
    super::IntrospectedElementKey, super::IntrospectionResult, super::Introspector, super::IntrospectorNavCommand,
    super::IsPacked, super::Packed, super::Removed, super::SavefileError, super::Schema, super::SchemaEnum,
    super::SchemaPrimitive, super::SchemaStruct, super::Serialize, super::Serializer, super::Variant,
    super::WithSchema, super::WithSchemaContext,
};

pub use byteorder::{LittleEndian, ReadBytesExt};
pub use memoffset::offset_of;
pub use memoffset::offset_of_tuple;
pub use memoffset::span_of;
pub use {super::AbiMethod, super::AbiMethodArgument, super::AbiMethodInfo, super::AbiTraitDefinition};

#[cfg(feature = "ring")]
pub use super::{load_encrypted_file, save_encrypted_file, CryptoReader, CryptoWriter};

#[cfg(feature = "derive")]
pub use savefile_derive::Packed;
#[cfg(feature = "derive")]
pub use savefile_derive::Savefile;
#[cfg(feature = "derive")]
pub use savefile_derive::SavefileIntrospectOnly;
#[cfg(feature = "derive")]
pub use savefile_derive::SavefileNoIntrospect;
