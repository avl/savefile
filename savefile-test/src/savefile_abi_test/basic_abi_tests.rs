use std::cell::{Cell, UnsafeCell};
use std::io::Cursor;

#[cfg(feature = "nightly")]
use test::Bencher;

use savefile::{AbiRemoved, Deserializer, Removed, Serialize, Serializer, ValueConstructor};
use savefile_abi::{abi_entry, AbiConnection, AbiExportable, AbiExportableImplementation};

#[savefile_abi_exportable(version = 0)]
pub trait CallbackInterface {
    fn set(&mut self, x: u32);
    fn get(&self) -> u32;
}

#[savefile_abi_exportable(version = 0)]
pub trait TestInterface {
    fn add(&self, x: u32, y: String) -> u32;
    fn call_callback(&mut self, callback: &mut dyn CallbackInterface);
    fn do_nothing(&self);
    fn do_panic(&self);
    fn arrays_add(&self, a: &[u32], b: &[u32]) -> Vec<u32>;
    fn string_arrays_add(&self, a: &[String], b: &[String]) -> Vec<String>;

    fn do_mut_nothing(&mut self);

    fn count_chars(&self, x: &String) -> usize;
    fn count_chars_str(&self, x: &str) -> usize;

    fn zero_sized_arg(&self, zero: ());
    fn simple_add(&self, a: u32, b: u32) -> u32;

    fn tuple_add1(&self, a: (u32,), b: (u32,)) -> (u32,);
    fn tuple_add2(&self, a: (u32, u32), b: (u32, u32)) -> (u32, u32);
    fn tuple_add3(&self, a: (u32, u32, u32), b: (u32, u32, u32)) -> (u32, u32, u32);

    fn boxes(&self, a: Box<u32>) -> Box<u32>;

    fn get_static_str(&self) -> &'static str;
}

#[derive(Default)]
pub struct TestInterfaceImpl {}

pub struct CallbackImpl {
    x: u32,
}

impl CallbackInterface for CallbackImpl {
    fn set(&mut self, x: u32) {
        self.x = x;
    }

    fn get(&self) -> u32 {
        self.x
    }
}

impl TestInterface for TestInterfaceImpl {
    fn add(&self, x: u32, y: String) -> u32 {
        x + y.parse::<u32>().unwrap()
    }

    fn call_callback(&mut self, callback: &mut dyn CallbackInterface) {
        callback.set(42);
    }
    fn arrays_add(&self, a: &[u32], b: &[u32]) -> Vec<u32> {
        let mut ret = Vec::new();
        for (a0, b0) in a.iter().copied().zip(b.iter().copied()) {
            ret.push(a0 + b0);
        }

        ret
    }
    fn do_nothing(&self) {}
    fn do_panic(&self) {
        panic!("TestInterface was asked to panic")
    }
    fn do_mut_nothing(&mut self) {}

    fn zero_sized_arg(&self, _zero: ()) {}

    fn simple_add(&self, a: u32, b: u32) -> u32 {
        a + b
    }

    fn tuple_add1(&self, a: (u32,), b: (u32,)) -> (u32,) {
        (a.0 + b.0,)
    }

    fn tuple_add2(&self, a: (u32, u32), b: (u32, u32)) -> (u32, u32) {
        (a.0 + b.0, a.1 + b.1)
    }

    fn tuple_add3(&self, a: (u32, u32, u32), b: (u32, u32, u32)) -> (u32, u32, u32) {
        (a.0 + b.0, a.1 + b.1, a.2 + b.2)
    }

    fn boxes(&self, a: Box<u32>) -> Box<u32> {
        a
    }

    fn string_arrays_add(&self, a: &[String], b: &[String]) -> Vec<String> {
        let mut ret = vec![];
        for (a1, b1) in a.iter().zip(b.iter()) {
            ret.push(a1.to_string() + b1);
        }
        ret
    }

    fn count_chars(&self, x: &String) -> usize {
        x.len()
    }
    fn count_chars_str(&self, x: &str) -> usize {
        x.len()
    }

    fn get_static_str(&self) -> &'static str {
        "hello world"
    }
}

savefile_abi_export!(TestInterfaceImpl, TestInterface);

#[test]
fn test_basic_call_abi() {
    let boxed: Box<dyn TestInterface> = Box::new(TestInterfaceImpl {});
    let mut conn = AbiConnection::from_boxed_trait(boxed).unwrap();

    let mut callback = CallbackImpl { x: 43 };
    conn.call_callback(&mut callback);

    assert_eq!(callback.x, 42);

    assert_eq!(conn.tuple_add1((1,), (2,)), (3,));
    assert_eq!(conn.tuple_add2((1, 1), (2, 2)), (3, 3));
    assert_eq!(conn.tuple_add3((1, 1, 1), (2, 2, 2)), (3, 3, 3));
    assert_eq!(conn.boxes(Box::new(42u32)), Box::new(42u32));

    assert_eq!(conn.count_chars(&"hejsan".to_string()), 6);
    assert_eq!(conn.count_chars_str("hejsan"), 6);
    assert_eq!(conn.get_static_str(), "hello world");
}

#[test]
fn test_slices() {
    let boxed: Box<dyn TestInterface> = Box::new(TestInterfaceImpl {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();

    let t = conn.arrays_add(&[1, 2, 3], &[1, 2, 3]);
    assert_eq!(t, vec![2, 4, 6]);

    let t = conn.string_arrays_add(&["hello ".to_string()], &["world".to_string()]);
    assert_eq!(t, vec!["hello world"]);
}

#[test]
fn test_zero_sized_arg() {
    let boxed: Box<dyn TestInterface> = Box::new(TestInterfaceImpl {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();
    conn.zero_sized_arg(());
}
#[test]
#[should_panic(expected = "TestInterface was asked to panic")]
fn test_panicking() {
    let boxed: Box<dyn TestInterface> = Box::new(TestInterfaceImpl {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();
    conn.do_panic();
}

#[test]
fn test_big_slices() {
    let boxed: Box<dyn TestInterface> = Box::new(TestInterfaceImpl {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();
    let a = vec![1u32; 10000];
    let b = vec![1u32; 10000];

    let t = conn.arrays_add(&a, &b);
    assert_eq!(t.len(), 10000);
    for x in t {
        assert_eq!(x, 2);
    }
}

struct FortyTwoConstructor {}
impl ValueConstructor<u32> for FortyTwoConstructor {
    fn make_value() -> u32 {
        42
    }
}

#[test]
fn test_abi_removed() {
    let removed: AbiRemoved<u32> = AbiRemoved::new();
    let mut data = Vec::new();
    Serializer::bare_serialize(&mut data, 0, &removed).unwrap();

    let roundtripped: u32 = Deserializer::bare_deserialize(&mut Cursor::new(&data), 0).unwrap();
    assert_eq!(roundtripped, 0);
}
#[test]
fn test_abi_removed_with_custom_default() {
    let removed: AbiRemoved<u32, FortyTwoConstructor> = AbiRemoved::<u32, FortyTwoConstructor>::new();
    let mut data = Vec::new();
    Serializer::bare_serialize(&mut data, 0, &removed).unwrap();

    let roundtripped: u32 = Deserializer::bare_deserialize(&mut Cursor::new(&data), 0).unwrap();
    assert_eq!(roundtripped, 42);
}

#[cfg(feature = "nightly")]
#[cfg(not(miri))]
#[bench]
fn bench_simple_call(b: &mut Bencher) {
    let boxed: Box<dyn TestInterface> = Box::new(TestInterfaceImpl {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();

    b.iter(move || conn.do_nothing())
}

#[cfg(feature = "nightly")]
#[cfg(not(miri))]
#[bench]
fn bench_simple_add(b: &mut Bencher) {
    let boxed: Box<dyn TestInterface> = Box::new(TestInterfaceImpl {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();

    b.iter(move || conn.simple_add(std::hint::black_box(1), std::hint::black_box(2)))
}
#[cfg(feature = "nightly")]
#[cfg(not(miri))]
#[bench]
fn bench_count_chars(b: &mut Bencher) {
    let boxed: Box<dyn TestInterface> = Box::new(TestInterfaceImpl {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();
    let mut s = String::new();
    use std::fmt::Write;
    for i in 0..10000 {
        write!(s, "{}", i).unwrap();
    }
    b.iter(move || conn.count_chars(std::hint::black_box(&s)))
}
#[cfg(feature = "nightly")]
#[cfg(not(miri))]
#[bench]
fn bench_count_chars_str(b: &mut Bencher) {
    let boxed: Box<dyn TestInterface> = Box::new(TestInterfaceImpl {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();
    let mut s = String::new();
    use std::fmt::Write;
    for i in 0..10000 {
        write!(s, "{}", i).unwrap();
    }
    b.iter(move || conn.count_chars_str(std::hint::black_box(&s)))
}
