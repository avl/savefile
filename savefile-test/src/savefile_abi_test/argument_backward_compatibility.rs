use savefile::prelude::AbiRemoved;
use savefile::{get_schema, SavefileError, WithSchemaContext};
use savefile_abi::RawAbiCallResult::AbiError;
use savefile_abi::{verify_compatiblity, AbiConnection, AbiExportable};
use savefile_abi_test::argument_backward_compatibility::v1::{ArgInterfaceV1, EnumArgument, Implementation1};
use savefile_abi_test::argument_backward_compatibility::v2::{ArgInterfaceV2, Implementation2};
use savefile_abi_test::basic_abi_tests::CowSmuggler;
use savefile_derive::Savefile;

mod v1 {
    #[derive(Savefile)]
    pub struct Argument {
        pub data1: u32,
        pub data2: u32,
    }

    #[derive(Savefile)]
    pub enum EnumArgument {
        Variant1,
        Variant2,
    }

    #[savefile_abi_exportable(version = 0)]
    pub trait ArgInterfaceV1 {
        fn sums(&self, a: Argument, b: Argument) -> u32;
        fn enum_arg(&self, a: EnumArgument) -> String;
        fn function_existing_in_v1(&self);
    }
    #[derive(Default)]
    pub struct Implementation1 {}

    impl ArgInterfaceV1 for Implementation1 {
        fn sums(&self, a: Argument, b: Argument) -> u32 {
            a.data1 + a.data2 + b.data1 + b.data2
        }
        fn enum_arg(&self, a: EnumArgument) -> String {
            match a {
                EnumArgument::Variant1 => "Variant1".into(),
                EnumArgument::Variant2 => "Variant2".into(),
            }
        }
        fn function_existing_in_v1(&self) {}
    }
}

mod v2 {
    use savefile::prelude::*;
    use savefile::AbiRemoved;
    use savefile_derive::Savefile;

    #[derive(Savefile, Debug)]
    pub struct ArgArgument {
        #[savefile_versions = "0..0"]
        pub data1: AbiRemoved<u32>,
        pub data2: u32,
        #[savefile_versions = "1.."]
        pub data3: u32,
    }
    #[derive(Savefile)]
    pub enum EnumArgument {
        Variant1,
        Variant2,
        #[savefile_versions = "1.."]
        Variant3,
    }

    #[savefile_abi_exportable(version = 1)]
    pub trait ArgInterfaceV2 {
        fn sums(&self, a: ArgArgument, b: ArgArgument) -> u32;
        fn enum_arg(&self, a: EnumArgument) -> String {
            match a {
                EnumArgument::Variant1 => "Variant1".into(),
                EnumArgument::Variant2 => "Variant2".into(),
                EnumArgument::Variant3 => "Variant3".into(),
            }
        }
        fn function_existing_in_v2(&self);
    }

    #[derive(Default)]
    pub struct Implementation2 {}
    impl ArgInterfaceV2 for Implementation2 {
        fn sums(&self, a: ArgArgument, b: ArgArgument) -> u32 {
            a.data3 + a.data2 + b.data2 + b.data3
        }

        fn function_existing_in_v2(&self) {}
    }
}

#[test]
#[cfg(not(miri))]
pub fn test_abi_schemas_get_def() {
    let exportable = <dyn ArgInterfaceV2 as AbiExportable>::get_definition(0);
    insta::assert_yaml_snapshot!(exportable);
}

#[test]
#[cfg(not(miri))]
pub fn test_backward_compatibility() -> Result<(), SavefileError> {
    verify_compatiblity::<dyn ArgInterfaceV2>("schemas")
}

#[test]
pub fn test_arg_argument_metadata() {
    use savefile::WithSchema;
    let schema = get_schema::<v2::ArgArgument>(0);
    println!("Schema: {:#?}", schema);
    assert!(!schema.layout_compatible(&schema)); //Versions containing removed items should never be considered layout compatible (since their schema type is not identical to the memory type)
}

#[test]
pub fn test_caller_has_older_version() {
    let iface2: Box<dyn ArgInterfaceV2> = Box::new(Implementation2 {});
    assert_eq!(
        iface2.sums(
            v2::ArgArgument {
                data2: 3,
                data3: 2,
                data1: AbiRemoved::new()
            },
            v2::ArgArgument {
                data2: 3,
                data3: 2,
                data1: AbiRemoved::new()
            }
        ),
        10
    );

    let conn1 = unsafe {
        AbiConnection::<dyn ArgInterfaceV1>::from_boxed_trait_for_test(
            <dyn ArgInterfaceV2 as AbiExportable>::ABI_ENTRY,
            iface2,
        )
    }
    .unwrap();

    let s = conn1.sums(v1::Argument { data1: 2, data2: 3 }, v1::Argument { data1: 4, data2: 5 });
    println!("Sum: {}", s);
    assert_eq!(s, 8); //Because implementation expects data2 and data3, but we're only sending data2.

    assert_eq!(conn1.enum_arg(EnumArgument::Variant1), "Variant1".to_string());
}

#[test]
pub fn test_caller_has_newer_version() {
    let iface1: Box<dyn ArgInterfaceV1> = Box::new(Implementation1 {});
    let conn1 = unsafe {
        AbiConnection::<dyn ArgInterfaceV2>::from_boxed_trait_for_test(
            <dyn ArgInterfaceV1 as AbiExportable>::ABI_ENTRY,
            iface1,
        )
    }
    .unwrap();

    assert_eq!(
        conn1.sums(
            v2::ArgArgument {
                data1: AbiRemoved::new(),
                data2: 1,
                data3: 2
            },
            v2::ArgArgument {
                data1: AbiRemoved::new(),
                data2: 3,
                data3: 4
            },
        ),
        4
    ); //Because implementation expects data1 and data2, but we're only sending data2.

    assert_eq!(conn1.enum_arg(v2::EnumArgument::Variant1), "Variant1".to_string());
}

#[test]
#[should_panic(expected = "Enum EnumArgument, variant Variant3 is not present in version 0")]
pub fn test_caller_has_newer_version_and_uses_enum_that_callee_doesnt_have() {
    let iface1: Box<dyn ArgInterfaceV1> = Box::new(Implementation1 {});
    let conn1 = unsafe {
        AbiConnection::<dyn ArgInterfaceV2>::from_boxed_trait_for_test(
            <dyn ArgInterfaceV1 as AbiExportable>::ABI_ENTRY,
            iface1,
        )
    }
    .unwrap();

    assert_eq!(conn1.enum_arg(v2::EnumArgument::Variant3), "Variant3".to_string());
}

#[test]
#[should_panic(expected = "'function_existing_in_v2' does not exist in implementation.")]
pub fn test_caller_has_newer_version_calling_non_existing_function() {
    let iface1: Box<dyn ArgInterfaceV1> = Box::new(Implementation1 {});
    let conn1 = unsafe {
        AbiConnection::<dyn ArgInterfaceV2>::from_boxed_trait_for_test(
            <dyn ArgInterfaceV1 as AbiExportable>::ABI_ENTRY,
            iface1,
        )
    }
    .unwrap();
    conn1.function_existing_in_v2();
}

#[test]
#[should_panic(expected = "'function_existing_in_v1' does not exist in implementation.")]
pub fn test_caller_has_older_version_calling_non_existing_function() {
    let iface2: Box<dyn ArgInterfaceV2> = Box::new(Implementation2 {});
    let conn = unsafe {
        AbiConnection::<dyn ArgInterfaceV1>::from_boxed_trait_for_test(
            <dyn ArgInterfaceV2 as AbiExportable>::ABI_ENTRY,
            iface2,
        )
    }
    .unwrap();
    conn.function_existing_in_v1();
}
#[test]
fn test_calling_function_that_is_later_removed() {
    let boxed: Box<dyn ArgInterfaceV1> = Box::new(Implementation1 {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();
    conn.function_existing_in_v1();
}
#[test]
fn test_calling_function_that_is_added_in_later_version() {
    let boxed: Box<dyn ArgInterfaceV2> = Box::new(Implementation2 {});
    let conn = AbiConnection::from_boxed_trait(boxed).unwrap();
    conn.function_existing_in_v2();
}
