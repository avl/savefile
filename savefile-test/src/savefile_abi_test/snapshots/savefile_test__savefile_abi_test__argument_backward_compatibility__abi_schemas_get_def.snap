---
source: savefile-test/src/savefile_abi_test/argument_backward_compatibility.rs
expression: exportable
---
name: ArgInterfaceV2
methods:
  - name: sums
    info:
      return_value:
        Primitive: schema_u32
      arguments:
        - schema:
            Struct:
              dbg_name: ArgArgument
              size: 8
              alignment: 4
              fields:
                - name: data1
                  value:
                    Primitive: schema_u32
                  offset: ~
                - name: data2
                  value:
                    Primitive: schema_u32
                  offset: 0
        - schema:
            Struct:
              dbg_name: ArgArgument
              size: 8
              alignment: 4
              fields:
                - name: data1
                  value:
                    Primitive: schema_u32
                  offset: ~
                - name: data2
                  value:
                    Primitive: schema_u32
                  offset: 0
  - name: enum_arg
    info:
      return_value:
        Primitive:
          schema_string: CapacityDataLength
      arguments:
        - schema:
            Enum:
              dbg_name: EnumArgument
              variants:
                - name: Variant1
                  discriminant: 0
                  fields: []
                - name: Variant2
                  discriminant: 1
                  fields: []
              discriminant_size: 1
              has_explicit_repr: false
              size: 1
              alignment: 1
  - name: function_existing_in_v2
    info:
      return_value: ZeroSize
      arguments: []
